gossipd/gossip_store.c

```diff
@@ -96,58 +97,69 @@ void gossip_store_add_channel_delete(struct gossip_store *gs,
 	tal_free(msg);
 }
 
-bool gossip_store_read_next(struct routing_state *rstate,
-			    struct gossip_store *gs)
+void gossip_store_load(struct routing_state *rstate, struct gossip_store *gs)
 {
 	beint32_t belen;
 	u32 msglen;
 	u8 *msg, *gossip_msg;
 	u64 satoshis;
-	enum gossip_wire_type type;
 	struct short_channel_id scid;
-
-	/* Did we already reach the end of the gossip_store? */
-	if (gs->read_pos == -1)
-		return NULL;
-
-	/* Can we read one message? */
-	if (pread(gs->fd, &belen, sizeof(belen), gs->read_pos) != sizeof(belen)) {
-		gs->read_pos = -1;
-		return false;
-	}
-
-	msglen = be32_to_cpu(belen);
-	msg = tal_arr(gs, u8, msglen);
-
-	if (!pread(gs->fd, msg, msglen, gs->read_pos + sizeof(belen))) {
-		status_trace("Short read from gossip-store, expected lenght %d",
-			     msglen);
-
-		/* Reset write_pos to truncate this message and disable future
-		 * reads */
-		gs->write_pos = gs->read_pos;
-		gs->read_pos = -1;
-		ftruncate(gs->fd, gs->write_pos);
+	/* We set/check version byte on creation */
+	off_t known_good = 1;
+	const char *bad;
+
+	lseek(gs->fd, known_good, SEEK_SET);
+	while (read(gs->fd, &belen, sizeof(belen)) == sizeof(belen)) {
+		msglen = be32_to_cpu(belen);
+		msg = tal_arr(gs, u8, msglen);
+
+		if (read(gs->fd, msg, msglen) != msglen) {
+			status_unusual("gossip_store: truncated file?");
+			goto truncate_nomsg;
+		}
+
+		if (fromwire_gossip_store_channel_announcement(msg, msg,
+							       &gossip_msg,
+							       &satoshis)) {
+			if (!routing_add_channel_announcement(rstate,
+							      gossip_msg,
+							      satoshis)) {
+				bad = "Bad channel_announcement";
+				goto truncate;
+			}
+		} else if (fromwire_gossip_store_channel_update(msg, msg,
+								&gossip_msg)) {
+			if (!routing_add_channel_update(rstate, gossip_msg)) {
+				bad = "Bad channel_update";
+				goto truncate;
+			}
+		} else if (fromwire_gossip_store_node_announcement(msg, msg,
+								   &gossip_msg)) {
+			if (!routing_add_node_announcement(rstate, gossip_msg)) {
+				bad = "Bad node_announcement";
+				goto truncate;
+			}
+		} else if (fromwire_gossip_store_channel_delete(msg, &scid)) {
+			struct chan *c = get_channel(rstate, &scid);
+			if (!c) {
+				bad = "Bad channel_delete";
+				goto truncate;
+			}
+			tal_free(c);
+		} else {
+			bad = "Unknown message";
+			goto truncate;
+		}
+		known_good += sizeof(belen) + msglen;
```

Comment:
We should probably just nuke the entire store. With this partial persistence we may end up adding channels that had a tombstone later in the store, giving us a myopic  view of the network as a whole.

Reply:
The store can get pretty large, so I'd rather not; I'm thinking of the gossip enhancements where we can efficiently ask for information past some timestamp.

Reply:
Ah,I see what you're saying.  We probably need to put block numbers into the gossip store, and if we truncate then we replay from that block?  For the moment, I'll truncate to zero.

Reply:
Going back to that block is also bad, because it means we might be syncing several thousand blocks. Truncating to zero is correct(-ish) IMHO, since this is ephemeral information that can be recovered by connecting to a peer.

Reply:
Yes, there has to be a limit.  In practice I think recent corruption is more likely, so it might be worth some heuristic in future.

Reply:
That sort of implies the order on the announcements matching the blockheight progression, which is not true, e.g., I could learn about a really old channel and then get a corruption, causing us to go back to that old block.